src/validator.js

```diff
@@ -0,0 +1,52 @@
+const { get, isNil } = require("lodash");
+const buildRoutingGraph = require("../src/helpers/build-routing-graph");
+const detectOrphan = require("./features/routing/detect-orphan");
+
+const failFast = errors => {
+  if (errors.errors.length > 0) {
+    throw new Error(JSON.stringify(errors, null, 2));
+  }
+};
+
+//eslint-disable-next-line
+const validatePage = (page, ctx, errors, options) => {
+  //Call your page-level validation scripts here
+  const graph = buildRoutingGraph(ctx);
+  const orphanError = detectOrphan(page.id, ctx, graph);
```

Comment:
Do we need to build this graph up every time we want to validate each page? Also had a thought, should orphan detection actually live in this page level validation function or should it be in the survey level validation as your detect-orphans module is validating the entire survey whenever it's called?

Reply:
We don't, no. It could be that the graph is built at the survey level and passed down; that's how it used to be. Changing it back to that probably makes sense, you're right to question it being built over and over as it's a waste of computation. I'll make that change.

I changed it originally so that it was built during the `detectOrphan()` function. But, this was then taken back out to make testing that function in isolation a little easier.

`detectOrpan()` isn't validating the entire survey when it's called. It requires the survey as a directed graph so that it can collect all of the in-edges to a page, from a given page id. It also requires the entire survey as a JSON string so that it can work out what the first question is, as this page will always have no in edges.

Although, those parameters could be changed such that `(id, ctx, graph)` becomes `id, firstQuestionID, inEdges)`... But, I might argue that as the `validator.js` module is where you **call** the validations, the logic to work out these values should exist within the module that uses them.

But, as it's not validating the entire survey I don't think it should live within `validateSurvey()`.